wasm-chain-integration/src/v1/trie/api.rs

```diff
@@ -0,0 +1,308 @@
+        match self {
+            PersistentState::Empty => None,
+            PersistentState::Root(node) => {
+                let data = node.use_value_(loader, |loader, node| node.data.lookup(loader, key))?;
+                let borrowed = data.borrow();
+                Some(borrowed.get(loader))
+            }
+        }
+    }
+
+    /// Generate a fresh mutable state from the persistent state.
+    pub fn thaw(&self) -> MutableState {
+        MutableState {
+            inner:      None,
+            persistent: self.clone(),
+        }
+    }
+
+    /// Compute the hash of the persistent state. This is a cheap operation
+    /// since the state hash is stored precomputed, however if the tree is not
+    /// loaded into memory then the hash will have to be retrieved from the
+    /// backing store using the provided loader.
+    pub fn hash(&self, loader: &mut impl BackingStoreLoad) -> super::Hash {
+        match self {
+            PersistentState::Empty => {
+                // hash of the node starts with either a 0 or 1 byte. This makes it distinct,
+                // but is otherwise an arbitrary choice.
+                super::Hash::from(<[u8; 32]>::from(sha2::Sha256::digest(b"empty contract state")))
+            }
+            PersistentState::Root(root) => root.hash(loader),
+        }
+    }
+
+    /// Cache the state, that is, load the entire state into memory from the
+    /// backing store. References to the backing store are retained.
+    pub fn cache<F: BackingStoreLoad>(&mut self, loader: &mut F) {
+        if let PersistentState::Root(node) = self {
+            node.load_and_cache(loader).data.cache(loader);
+        }
+    }
+
+    #[cfg(feature = "display-state")]
+    pub fn display_tree(&self, builder: &mut TreeBuilder, loader: &mut impl BackingStoreLoad) {
+        match self {
+            Self::Empty => {}
+            Self::Root(node) => {
+                node.use_value_(loader, |loader, tree| tree.data.display_tree(builder, loader))
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// This type is a technical device to support lazy conversion of
+/// [PersistentState] to [MutableState]. It contains the runtime representation
+/// of the tree that supports efficient updates.
+/// Clone on this structure is designed to be cheap, **and it is a shallow
+/// copy**. Modifications of the inner MutableTrie on either the clone or the
+/// original propagate to the other.
+pub struct MutableStateInner {
+    /// Current root of the tree. The current generation.
+    root:  u32,
+    /// The mutable trie itself. The root is an index in the array of generation
+    /// roots.
+    /// The idea is that the mutex is acquired at the start of execution of the
+    /// contract and released at the end.
+    /// The reason for the mutex is that we need to be able to clone this so
+    /// that we can share it inside the single transaction.
+    state: Arc<Mutex<MutableTrie>>,
```

Comment:
We need to be able to mutate the state, so we need the mutex. If we just had `Arc<MutableTrie>` then it is never possible to get a mutable reference to `MutableTrie`, and we need that.

`Arc<Mutex<T>>` is the equivalent to `Rc<RefCell<T>>`.

While it is true that execution is single-threaded the general execution context is multithreaded since (in general at least) we are using a parallel garbage collector. 

In this specific case I think it might be possible to use `Rc<RefCell<T>>` instead, but we'd need to carefully analyze that we don't have issues with cleanup (which is triggered via the Haskell GC in some cases). Since this does not impact performance I don't think the risks and time spent analyzing to make sure it would work are worth it.

We need multiple copies of this due to generations and rollback.